### PR TITLE
#438 - Fix group_by for DataFrame with single row

### PR DESCRIPTION
**Pull request by baarkerlounger**
https://github.com/SciRuby/daru/issues/438

### CODE REVIEW

File: spec/dataframe_spec.rb [hunk 1]
```diff
@@ -3266,6 +3266,15 @@
     end
   end
 
+  context "group_by" do
+    context "on a single row DataFrame" do
+      let(:df){ Daru::DataFrame.new({a: [1], b: [2]}) }
+      it "returns itself" do
```
Comment (zverok) [inline spec/dataframe_spec.rb]: I am not sure it is the most reasonable behavior. Here is what happens if we have large DF:
```ruby
df = Daru::DataFrame.new(city: %w[Kyiv Kharkiv Odesa Kyiv Kharkiv Odesa], year: [2015, 2015, 2015, 2016, 2016, 2016], value: [1, 2, 3, 4, 5, 6])
# => #<Daru::DataFrame(6x3)>
#            city    year   value
#       0    Kyiv    2015       1
#       1 Kharkiv    2015       2
#       2   Odesa    2015       3
#       3    Kyiv    2016       4
#       4 Kharkiv    2016       5
#       5   Odesa    2016       6 
df.group_by(:city)
# => #<Daru::DataFrame(6x2)>
#                    year   value
# Kharkiv       1    2015       2
#               4    2016       5
#    Kyiv       0    2015       1
#               3    2016       4
#   Odesa       2    2015       3
#               5    2016       6 
```
Here is what (I believe) should happen with 1-row DF:
```ruby
df = Daru::DataFrame.new(city: %w[Kyiv], year: [2015], value: [1])
# => #<Daru::DataFrame(1x3)>
#        city  year value
#     0  Kyiv  2015     1 
df.group_by(:city)
# => #<Daru::DataFrame(1x2)>
#                    year   value
#    Kyiv       0    2015       1
```
(E.g. the same index/vectors change as with other groupings).

File: lib/daru/dataframe.rb [hunk 2]
```diff
@@ -1449,12 +1449,13 @@ def vector_mean max_missing=0
...
+      return self if nrows == 1
       vectors.flatten!
-      # FIXME: wouldn't it better to do vectors - @vectors here and
-      # raise one error with all non-existent vector names?.. - zverok, 2016-05-18
-      vectors.each { |v|
-        raise(ArgumentError, "Vector #{v} does not exist") unless has_vector?(v)
-      }
+      missing = vectors - @vectors.to_a
+      unless missing.empty?
+        e = missing.size > 1 ? "Vectors #{missing.join(', ')} do not exist" : "Vector #{missing.first} does not exist"
```
[the remark this answers is not in the document]
Comment (baarkerlounger) [reply lib/daru/dataframe.rb]: Updated

File: spec/dataframe_spec.rb (hunk 1 above)
Comment (baarkerlounger) [reply spec/dataframe_spec.rb]: Fair point, made it match the behaviour of other cases.

File: lib/daru/dataframe.rb [hunk 3]
```diff
@@ -173,7 +173,8 @@ def rows source, opts={}
         elsif ArrayHelper.array_of?(source, Vector)
           from_vector_rows(source, opts)
         else
-          raise ArgumentError, "Can't create DataFrame from #{source}"
+          row = source.map { |v| Array(v) }
```
Comment (zverok) [inline lib/daru/dataframe.rb]: What would be consequences? What if `source` doesn't respond to `map`? 

And what if it _does_ respond, where are the tests for the change? Now not only `group_by` behavior affected.

Comment (baarkerlounger) [reply lib/daru/dataframe.rb]: Have added a check for that case.

In terms of specs - I've added one for the breaking case and all others are still passing, do you have a particular spec case in mind?

Comment (zverok) [reply lib/daru/dataframe.rb]: Not that. Rather, I believe that local bug should have local fix, e.g. either in `GroupBy#resultant_context` (where DF is tried to be created from "wrong" data), or in `DataFrame#access_row_tuples_by_indexs` (which probably should always return array-of-arrays, shouldn't it?)

Comment (baarkerlounger) [reply lib/daru/dataframe.rb]: Ah I get it, agreed that's a better way. Have updated.

**Comment by baarkerlounger** [comment]
@zverok happy to merge?

**Comment by baarkerlounger** [comment]
@zverok updated

**Comment by zverok** [comment]
Thanks!

**Pull request merged**